### ISSUE TITLE
Incorrect cell size with FreeType fonts containing SVG glyphs

### ISSUE THREAD

**Issue by tokyovigilante** [alacritty/alacritty#6048]
I'm running a mixed-DPI system on Wayland, and since a recent system upgrade have noticed that Alacritty's cell size and scaling is incorrect, leading to rendering issues.

I note the cell size has a row height of 1 even after scaling calculation. I also not an erroneous scale factor of 2.0 even though the display has a custom scale of 1.6 set.

### System

OS: Linux 5.18-rc5, Sway/Wayland (git head)
Version: Version 0.11.0-dev (ad3ebba5)

### Logs

`alacritty -vv`
```
[0.000000513s] [INFO ] [alacritty] Welcome to Alacritty
[0.000020223s] [INFO ] [alacritty] Version 0.11.0-dev (ad3ebba5)
[0.000030320s] [INFO ] [alacritty] Running on Wayland
[0.000350593s] [INFO ] [alacritty] Configuration files loaded from:
                                     "/home/ryan/.config/alacritty/alacritty.yml"
[0.000450505s] [DEBUG] [alacritty] Loading "Source Code Pro" font
[0.003180789s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.003799465s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.004404253s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.005029226s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.005298649s] [DEBUG] [alacritty] Estimated scaling factor: 1
[0.005303348s] [DEBUG] [alacritty] Estimated window size: None
[0.005305826s] [DEBUG] [alacritty] Estimated cell size: 9 x 1
[0.023991131s] [INFO ] [alacritty] Running on AMD Radeon VII (vega20, LLVM 14.0.0, DRM 3.46, 5.18.0-rc5-crackotage+)
[0.024007439s] [INFO ] [alacritty] Using OpenGL 3.3 renderer
[0.026859802s] [INFO ] [alacritty] Display scale factor: 1
[0.026870120s] [DEBUG] [alacritty] Filling glyph cache with common glyphs
[0.032727780s] [INFO ] [alacritty] Cell size: 9 x 1
[0.032743722s] [INFO ] [alacritty] Padding: 0 x 0
[0.032746229s] [INFO ] [alacritty] Width: 800, Height: 600
[0.033171900s] [INFO ] [alacritty] PTY dimensions: 600 x 88
[0.037461805s] [INFO ] [alacritty] Initialisation complete
[0.045797019s] [DEBUG] [alacritty_terminal] New num_cols is 141 and num_lines is 1313
[0.049704101s] [INFO ] [alacritty] Padding: 0 x 0
[0.049720087s] [INFO ] [alacritty] Width: 1272, Height: 1313
[0.060586549s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.061197837s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.061779894s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.062400138s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.062679153s] [INFO ] [alacritty] Font size changed to Size(21) with scale factor of 2
[0.065081724s] [INFO ] [alacritty] Cell size: 9 x 1
[0.065093098s] [DEBUG] [alacritty_terminal] New num_cols is 282 and num_lines is 2626
[0.074493851s] [INFO ] [alacritty] Padding: 0 x 0
[0.074508658s] [INFO ] [alacritty] Width: 2544, Height: 2626
```

`swaymsg -t get_outputs`
```
Output DP-1 'Dell Inc. DELL P2415Q D3C8X4C50E8L'
  Current mode: 3840x2160 @ 59.997 Hz
  Position: 0,270
  Scale factor: 2.000000
  Scale filter: nearest
  Subpixel hinting: unknown
  Transform: normal
  Workspace: 1
  Max render time: 3 ms
  Adaptive sync: disabled
  Available modes:
    3840x2160 @ 59.997 Hz
    2560x1440 @ 59.951 Hz
    1920x1200 @ 59.885 Hz
    1920x1080 @ 60.000 Hz
    1920x1080 @ 60.000 Hz
    1920x1080 @ 59.940 Hz
    1920x1080 @ 50.000 Hz
    1920x1080 @ 30.000 Hz
    1920x1080 @ 29.970 Hz
    1920x1080 @ 25.000 Hz
    1920x1080 @ 24.000 Hz
    1920x1080 @ 23.976 Hz
    1600x1200 @ 60.000 Hz
    1680x1050 @ 59.997 Hz
    1600x900 @ 60.000 Hz
    1280x1024 @ 75.025 Hz
    1280x1024 @ 60.020 Hz
    1440x900 @ 59.997 Hz
    1280x800 @ 59.997 Hz
    1152x864 @ 75.000 Hz
    1280x720 @ 60.000 Hz
    1280x720 @ 59.940 Hz
    1280x720 @ 50.000 Hz
    1024x768 @ 75.029 Hz
    1024x768 @ 60.004 Hz
    800x600 @ 75.000 Hz
    800x600 @ 60.317 Hz
    720x576 @ 50.000 Hz
    720x480 @ 60.000 Hz
    720x480 @ 60.000 Hz
    720x480 @ 59.940 Hz
    720x480 @ 59.940 Hz
    640x480 @ 75.000 Hz
    640x480 @ 60.000 Hz
    640x480 @ 59.940 Hz
    640x480 @ 59.940 Hz
    720x400 @ 70.082 Hz

Output DP-2 'Eizo Nanao Corporation MX315W 0x00009D25' (focused)
  Current mode: 4096x2160 @ 59.983 Hz
  Position: 1920,0
  Scale factor: 1.600000
  Scale filter: linear
  Subpixel hinting: unknown
  Transform: normal
  Workspace: 3
  Max render time: 3 ms
  Adaptive sync: disabled
  Available modes:
    4096x2160 @ 59.983 Hz
    3840x2160 @ 59.997 Hz
    1920x1200 @ 59.983 Hz
    1920x1080 @ 59.983 Hz
    1600x1200 @ 60.000 Hz
    1680x1050 @ 59.983 Hz
    1280x1024 @ 60.020 Hz
    1440x900 @ 59.983 Hz
    1280x800 @ 59.983 Hz
    1280x720 @ 59.983 Hz
    1024x768 @ 60.004 Hz
    800x600 @ 60.317 Hz
    640x480 @ 59.940 Hz
    720x400 @ 70.082 Hz
```

**Comment by chrisduerr** [comment]
> I also not an erroneous scale factor of 2.0 even though the display has a custom scale of 1.6 set.

Even though Alacritty would support a fractional scale factor, those are actually never reported to the application. Instead the nearest integer factor is used and scaling is done from there.

> since a recent system upgrade

Did you upgrade Alacritty? If so, does a previous version of Alacritty work properly?

**Comment by tokyovigilante** [comment]
Thanks for the response.The same behaviour was seen before and after pulling from master and rebuilding. It is also seen on my other display which is using an integer scale factor.

For reference I am running Fedora rawhide so this was a `dnf upgrade`after a month away. Initially I thought it was a video driver bug but no other application is affected.

**Comment by marduone** [comment]
Seems to happen on X11 too, if by rendering issues it is meant that the terminal output is garbled. 

I updated my system (openSUSE Tumbleweed) today with `zypper dup` as is customary with that distro. Alacritty package version did not change, so I suspect the problem is caused by some other package / library that was updated.

I first noticed the problem in Sway/Wayland, like OP, where alacritty is set as default terminal but the same problem is also present in KWin/Wayland, KWin/X11, i3/X11, and icewm/X11.

### System

openSUSE Tumbleweed
kernel 5.17.4
alacritty-0.10.1-2.1.x86_64

RPM package version says alacritty-0.10.1-2.1.x86_64 but it must have been built from master some time after 0.10.1 was released, because the CHANGELOG.md included in the package has these lines:

```
## 0.11.0-dev

### Packaging

- Minimum Rust version has been bumped to 1.56.0

### Added

- Track and report surface damage information to Wayland compositors
- Escape sequence for undercurl, dotted and dashed underlines (`CSI 4 : [3-5] m`)

### Changed

- The `--help` output was reworked with a new colorful syntax

### Fixed

- Creating the IPC socket failing if WAYLAND_DISPLAY contains an absolute path

## 0.10.1
```

I've no idea what the actual commit id is/was.

### Logs

Sway/Wayland:

```
[0.000009689s] [INFO ] [alacritty] Welcome to Alacritty
[0.000281966s] [INFO ] [alacritty] Version 0.11.0-dev
[0.000330578s] [INFO ] [alacritty] Running on Wayland
[0.001122331s] [INFO ] [alacritty] Configuration files loaded from:
                                     "/home/mlaakson/.config/alacritty/alacritty.yml"
[0.001954350s] [DEBUG] [alacritty] Loading "monospace" font
[0.048091872s] [TRACE] [crossfont] Number of fonts is 58
[0.048285309s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-Regular.otf", index=0
[0.048416668s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.061547794s] [TRACE] [crossfont] Number of fonts is 59
[0.061743927s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-Bold.otf", index=0
[0.061867622s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.077832809s] [TRACE] [crossfont] Number of fonts is 58
[0.078049480s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-It.otf", index=0
[0.078197150s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.091962941s] [TRACE] [crossfont] Number of fonts is 58
[0.092139917s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-BoldIt.otf", index=0
[0.092264233s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.098299420s] [DEBUG] [alacritty] Estimated scaling factor: 1
[0.098315350s] [DEBUG] [alacritty] Estimated window size: None
[0.098321852s] [DEBUG] [alacritty] Estimated cell size: 9 x 1
[0.138403859s] [INFO ] [alacritty] Display scale factor: 1
[0.138448845s] [DEBUG] [alacritty] Filling glyph cache with common glyphs
[0.149928266s] [INFO ] [alacritty] Cell size: 9 x 1
[0.149978511s] [INFO ] [alacritty] Padding: 0 x 0
[0.149988159s] [INFO ] [alacritty] Width: 800, Height: 600
[0.153865820s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 197, g: 200, b: 198 } to Rgb { r: 130, g: 132, b: 130 }
[0.153885407s] [TRACE] [alacritty] Deriving dim colors from normal colors
[0.153891369s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 29, g: 31, b: 33 } to Rgb { r: 19, g: 20, b: 21 }
[0.153897861s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 204, g: 102, b: 102 } to Rgb { r: 134, g: 67, b: 67 }
[0.153903913s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 181, g: 189, b: 104 } to Rgb { r: 119, g: 124, b: 68 }
[0.153909573s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 240, g: 198, b: 116 } to Rgb { r: 158, g: 130, b: 76 }
[0.153915234s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 129, g: 162, b: 190 } to Rgb { r: 85, g: 106, b: 125 }
[0.153920975s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 178, g: 148, b: 187 } to Rgb { r: 117, g: 97, b: 123 }
[0.153928960s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 138, g: 190, b: 183 } to Rgb { r: 91, g: 125, b: 120 }
[0.153934861s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 197, g: 200, b: 198 } to Rgb { r: 130, g: 132, b: 130 }
[0.153942426s] [INFO ] [alacritty] PTY dimensions: 600 x 88
[0.157238743s] [INFO ] [alacritty] Initialisation complete
[0.168107665s] [DEBUG] [alacritty_terminal] New num_cols is 141 and num_lines is 1379
[0.179349034s] [INFO ] [alacritty] Padding: 0 x 0
[0.179413346s] [INFO ] [alacritty] Width: 1276, Height: 1379
```

i3/X11:

```
[0.000001853s] [INFO ] [alacritty] Welcome to Alacritty
[0.000142380s] [INFO ] [alacritty] Version 0.11.0-dev
[0.000165724s] [INFO ] [alacritty] Running on X11
[0.000687285s] [INFO ] [alacritty] Configuration files loaded from:
                                     "/home/mlaakson/.config/alacritty/alacritty.yml"
[0.002610105s] [DEBUG] [alacritty] Loading "monospace" font
[0.026628995s] [TRACE] [crossfont] Number of fonts is 57
[0.026829736s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-Regular.otf", index=0
[0.026967467s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.042980568s] [TRACE] [crossfont] Number of fonts is 58
[0.043176370s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-Bold.otf", index=0
[0.043306748s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.057310354s] [TRACE] [crossfont] Number of fonts is 57
[0.057498301s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-It.otf", index=0
[0.057632014s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.073015310s] [TRACE] [crossfont] Number of fonts is 57
[0.073231420s] [TRACE] [crossfont] Got font path="/usr/share/fonts/truetype/SourceCodePro-BoldIt.otf", index=0
[0.073360234s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", lcd_filter: 1 }
[0.080152900s] [DEBUG] [alacritty] Estimated scaling factor: 1.1666666666666667
[0.080185242s] [DEBUG] [alacritty] Estimated window size: None
[0.080192606s] [DEBUG] [alacritty] Estimated cell size: 9 x 1
[0.140466406s] [INFO ] [alacritty] Display scale factor: 1.1666666666666667
[0.140508455s] [DEBUG] [alacritty] Filling glyph cache with common glyphs
[0.153127313s] [INFO ] [alacritty] Cell size: 9 x 1
[0.153164503s] [INFO ] [alacritty] Padding: 0 x 0
[0.153171186s] [INFO ] [alacritty] Width: 800, Height: 600
[0.170715865s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 197, g: 200, b: 198 } to Rgb { r: 130, g: 132, b: 130 }
[0.170758776s] [TRACE] [alacritty] Deriving dim colors from normal colors
[0.170767002s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 29, g: 31, b: 33 } to Rgb { r: 19, g: 20, b: 21 }
[0.170776620s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 204, g: 102, b: 102 } to Rgb { r: 134, g: 67, b: 67 }
[0.170785637s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 181, g: 189, b: 104 } to Rgb { r: 119, g: 124, b: 68 }
[0.170793863s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 240, g: 198, b: 116 } to Rgb { r: 158, g: 130, b: 76 }
[0.170800826s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 129, g: 162, b: 190 } to Rgb { r: 85, g: 106, b: 125 }
[0.170808010s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 178, g: 148, b: 187 } to Rgb { r: 117, g: 97, b: 123 }
[0.170814953s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 138, g: 190, b: 183 } to Rgb { r: 91, g: 125, b: 120 }
[0.170821766s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 197, g: 200, b: 198 } to Rgb { r: 130, g: 132, b: 130 }
[0.170832246s] [INFO ] [alacritty] PTY dimensions: 600 x 88
[0.176012530s] [INFO ] [alacritty] Initialisation complete
[0.182090980s] [DEBUG] [alacritty_terminal] New num_cols is 141 and num_lines is 1391
[0.189292743s] [INFO ] [alacritty] Padding: 0 x 0
[0.189347567s] [INFO ] [alacritty] Width: 1276, Height: 1391
```

**Comment by chrisduerr** [comment]
Which font are you using? Does changing the font fix this?

**Comment by tokyovigilante** [comment]
> Which font are you using? Does changing the font fix this?

Interestingly yes, usually Source Code Pro, however with B612 Mono:

```
[0.000001895s] [INFO ] [alacritty] Welcome to Alacritty
[0.000067506s] [INFO ] [alacritty] Version 0.11.0-dev (ad3ebba5)
[0.000095896s] [INFO ] [alacritty] Running on Wayland
[0.001323091s] [INFO ] [alacritty] Configuration files loaded from:
                                     "/home/ryan/.config/alacritty/alacritty.yml"
[0.001708593s] [DEBUG] [alacritty] Loading "B612 Mono" font
[0.007868008s] [TRACE] [crossfont] Number of fonts is 37
[0.007892638s] [TRACE] [crossfont] Got font path="/home/ryan/.fonts/B612Mono-Regular.ttf", index=0
[0.007933278s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.008738502s] [TRACE] [crossfont] Number of fonts is 38
[0.008754485s] [TRACE] [crossfont] Got font path="/home/ryan/.fonts/B612Mono-Bold.ttf", index=0
[0.008777451s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.009504690s] [TRACE] [crossfont] Number of fonts is 41
[0.009516864s] [TRACE] [crossfont] Got font path="/home/ryan/.fonts/B612Mono-Italic.ttf", index=0
[0.009537164s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.010189500s] [TRACE] [crossfont] Number of fonts is 42
[0.010201440s] [TRACE] [crossfont] Got font path="/home/ryan/.fonts/B612Mono-BoldItalic.ttf", index=0
[0.010220432s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT, render_mode: "Normal", lcd_filter: 1 }
[0.010636365s] [DEBUG] [alacritty] Estimated scaling factor: 1
[0.010641074s] [DEBUG] [alacritty] Estimated window size: None
[0.010643213s] [DEBUG] [alacritty] Estimated cell size: 9 x 18
[0.029469291s] [INFO ] [alacritty] Running on AMD Radeon VII (vega20, LLVM 14.0.0, DRM 3.46, 5.18.0-rc5-crackotage+)
[0.029479502s] [INFO ] [alacritty] Using OpenGL 3.3 renderer
[0.032098258s] [INFO ] [alacritty] Display scale factor: 1
[0.032106800s] [DEBUG] [alacritty] Filling glyph cache with common glyphs
[0.038174810s] [INFO ] [alacritty] Cell size: 9 x 18
```

**Comment by Studiedlist** [comment]
I think, the problem is caused by [crossfont](https://docs.rs/crossfont/latest/crossfont/index.html) crate, which calculates line height wrong for some fonts (line height is 0 at [compute_cell_size](https://github.com/alacritty/alacritty/blob/5c51d528ea8ded63b127a4f17ee6136653db09b9/alacritty/src/display/mod.rs#L1147) function)
Since the error appears not to be caused by alacritty update, I think something has been changed in system libraries and caused crossfont work wrong

**Comment by solacelost** [comment]
I am seeing identical behavior to the above after updating Fedora 35 with system alacritty to Fedora 36 (which should come out of beta officially on Tuesday). I also use Source Code Pro as my main font, which is in version 2.030.1.050 in both Fedora 35 and 36 [1]. Alacritty is at version 0.10.1 for both as well [2], confirming that it's likely that some other change in the system is to blame. Hopefully that helps narrow it down.

1. https://koji.fedoraproject.org/koji/packageinfo?packageID=17006
2. https://koji.fedoraproject.org/koji/packageinfo?packageID=32507

**Comment by shunghsiyu** [comment]
This seems to happen when **libfreetype** is updated to `v2.12.1`, it was fine previously with `v2.12.0`.

I've also encountered the same issue on the latest openSUSE Tumbleweed update, but if I lock the version of libfreetype to `v2.12.0`, the issue do not occur.

**Comment by lseelenbinder** [comment]
I can confirm the same happening (with the same font) on Fedora 36.

**Comment by chrisduerr** [comment]
> This seems to happen when libfreetype is updated to v2.12.1, it was fine previously with v2.12.0.

That's interesting, seems like FreeType might have changed things beneath us again. While they have dropped support for things they considered "incorrect usage" in minor/patch versions before, I feel like font metrics should be simple enough that I doubt we're doing something crazy here. So it's entirely possible this is FreeType's/SCP's fault.

I don't see any relevant changes in [the changelog](https://gitlab.freedesktop.org/freetype/freetype/-/blob/master/docs/CHANGES#L1-26) at least. And on Archlinux I cannot reproduce this either.

Can anyone reproduce this on a distro other than Fedora/openSUSE?

**Comment by rudi10-1** [comment]
This happens on Debian testing. Source Code Pro results in cell size: 9 x 1. Changing font to Fira Code fixed the issue. This came up after Friday upgrade of libfreetype from 2.11.1 to 2.12.0.

**Comment by chrisduerr** [comment]
> This seems to happen when libfreetype is updated to v2.12.1, it was fine previously with v2.12.0.

> This came up after Friday upgrade of libfreetype from 2.11.1 to 2.12.0.

It seems like there's some mixed messages here. @shunghsiyu can you confirm that this wasn't a typo and you got the breakage on a different update than @rudi10-1?

Getting breakage on the 2.11.1 -> 2.12.0 transition seems much more likely, since that changed much more stuff especially wrt the SVGs. Especially since IIRC the SVG emoji stuff is something introduced by adobe so I wouldn't be surprised if SCP made use of that.

**Comment by tokyovigilante** [comment]
My system went from 2.11.1 to 2.12.1 directly, and agree there is very little churn between 2.12.0 to 2.12.1, so I'd wager the issue was introduced between 2.11.1 to 2.12.0. Regrettably the 2.11.1 package is gone from rawhide so I can't easily roll back to it.

**Comment by chrisduerr** [comment]
> Regrettably the 2.11.1 package is gone from rawhide so I can't easily roll back to it.

That's very unfortunate. I still find it strange that I cannot reproduce this on Archlinux. Could you tell me what kind of files are installed on Fedora/Debian for this font? On Archlinux it seems to be a couple OTF fonts:

```
adobe-source-code-pro-fonts /usr/
adobe-source-code-pro-fonts /usr/share/
adobe-source-code-pro-fonts /usr/share/fonts/
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-Black.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-BlackIt.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-Bold.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-BoldIt.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-ExtraLight.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-ExtraLightIt.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-It.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-Light.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-LightIt.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-Medium.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-MediumIt.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-Regular.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-Semibold.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodePro-SemiboldIt.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodeVariable-Italic.otf
adobe-source-code-pro-fonts /usr/share/fonts/adobe-source-code-pro/SourceCodeVariable-Roman.otf
adobe-source-code-pro-fonts /usr/share/licenses/
adobe-source-code-pro-fonts /usr/share/licenses/adobe-source-code-pro-fonts/
adobe-source-code-pro-fonts /usr/share/licenses/adobe-source-code-pro-fonts/LICENSE.md
```

So if those are `ttf` fonts on Fedora/Debian that would explain why it works on Archlinux.

**Comment by jfkw** [comment]
I can confirm same behavior (fonts squashed vertically, about 1px per text line sounds about right) on openSUSE Tumbleweed 20220516 with the following package versions:

```
$ rpm -q alacritty libfreetype6 adobe-sourcecodepro-fonts
alacritty-0.10.1-59.3.x86_64
libfreetype6-2.12.1-1.2.x86_64
adobe-sourcecodepro-fonts-2.030-3.7.noarch
```

Using awesomewm X11 tiling window manager, alacritty log as follows:

```
$ alacritty -vvv
Created log file at "/tmp/Alacritty-16039.log"
[0.000004311s] [INFO ] [alacritty] Welcome to Alacritty
[0.000036101s] [INFO ] [alacritty] Version 0.11.0-dev
[0.000048469s] [INFO ] [alacritty] Running on X11
[0.000168859s] [INFO ] [alacritty_config_derive] No config file found; using default
[0.006416312s] [DEBUG] [alacritty] Loading "monospace" font
[0.044260207s] [TRACE] [crossfont] Number of fonts is 254
[0.044419256s] [TRACE] [crossfont] Got font path="/usr/share/fonts/texlive-sourcecodepro/SourceCodePro-Regular.otf", index=0
[0.044556895s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd"
, lcd_filter: 1 }
[0.079787858s] [TRACE] [crossfont] Number of fonts is 265
[0.079933827s] [TRACE] [crossfont] Got font path="/usr/share/fonts/texlive-sourcecodepro/SourceCodePro-Bold.otf", index=0
[0.080051264s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd", l
cd_filter: 1 }
[0.114070589s] [TRACE] [crossfont] Number of fonts is 267
[0.114232018s] [TRACE] [crossfont] Got font path="/usr/share/fonts/texlive-sourcecodepro/SourceCodePro-RegularIt.otf", index=0
[0.114348094s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Italic, load_flags: TARGET_LIGHT | COLOR, render_mode: "Lcd",
 lcd_filter: 1 }
[0.148549523s] [TRACE] [crossfont] Number of fonts is 280
[0.148705562s] [TRACE] [crossfont] Got font path="/usr/share/fonts/texlive-sourcecodepro/SourceCodePro-BoldIt.otf", index=0
[0.148823592s] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold Italic, load_flags: TARGET_LIGHT | COLOR, render_mode: "
Lcd", lcd_filter: 1 }
[0.176740243s] [DEBUG] [alacritty] Estimated scaling factor: 1.5
[0.176763557s] [DEBUG] [alacritty] Estimated window size: None
[0.176768897s] [DEBUG] [alacritty] Estimated cell size: 9 x 1
[0.253127691s] [INFO ] [alacritty] Display scale factor: 1.5
[0.253150069s] [DEBUG] [alacritty] Filling glyph cache with common glyphs
[0.263808239s] [INFO ] [alacritty] Cell size: 9 x 1
[0.263827613s] [INFO ] [alacritty] Padding: 0 x 0
[0.263833782s] [INFO ] [alacritty] Width: 800, Height: 600
[0.270223577s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 197, g: 200, b: 198 } to Rgb { r: 130, g: 132, b: 130 }
[0.270239737s] [TRACE] [alacritty] Deriving dim colors from normal colors
[0.270245301s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 29, g: 31, b: 33 } to Rgb { r: 19, g: 20, b: 21 }
[0.270250801s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 204, g: 102, b: 102 } to Rgb { r: 134, g: 67, b: 67 }
[0.270255618s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 181, g: 189, b: 104 } to Rgb { r: 119, g: 124, b: 68 }
[0.270260359s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 240, g: 198, b: 116 } to Rgb { r: 158, g: 130, b: 76 }
[0.270265178s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 129, g: 162, b: 190 } to Rgb { r: 85, g: 106, b: 125 }
[0.270270709s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 178, g: 148, b: 187 } to Rgb { r: 117, g: 97, b: 123 }
[0.270276281s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 138, g: 190, b: 183 } to Rgb { r: 91, g: 125, b: 120 }
[0.270281788s] [TRACE] [alacritty_terminal] Scaling RGB by 0.66 from Rgb { r: 197, g: 200, b: 198 } to Rgb { r: 130, g: 132, b: 130 }
[0.270293652s] [INFO ] [alacritty] PTY dimensions: 600 x 88
[0.273828433s] [INFO ] [alacritty] Initialisation complete
[0.287861891s] [DEBUG] [alacritty_terminal] New num_cols is 213 and num_lines is 1055
[0.294502284s] [INFO ] [alacritty] Padding: 0 x 0
[0.294526870s] [INFO ] [alacritty] Width: 1918, Height: 1055
[0.547323316s] [TRACE] [alacritty_terminal] Saving cursor position
[0.547344334s] [TRACE] [alacritty_terminal] Setting title to 'Some("user@host:~")'
[0.547379481s] [TRACE] [alacritty_terminal] Restoring cursor position
[32.476368384s] [INFO ] [alacritty] Goodbye
```

**Comment by chrisduerr** [comment]
Hm, Archlinux has `2.038ro+1.058it+1.018var-1` of SCP, so maybe an update to the SCP font is necessary to fix this?

**Comment by chrisduerr** [comment]
I've uploaded Archlinux's version of SCP, would be curious to see if manually replacing it solves the problem. You'll have to replace the `.jpeg` with `.otf`, unfortunately GitHub doesn't accept arbitrary files.

![SourceCodePro-Regular](https://user-images.githubusercontent.com/8886672/168931730-8b176a50-e4d9-40b2-bab6-67906dd463c8.jpeg)

Important to note that you'll have to refresh your fontconfig cache manually when replacing fonts by hand (see `fc-cache`).

**Comment by tokyovigilante** [comment]
I took a moment to manually install freetype from git, then rolled back to 2.11 and this fixed the issue, so it's fair to say a recent FT change is to blame. 

`git bisect` reveals:
```
[Wed 12:14PM] ryan@crackotage ~/Projects/Develop/freetype/$ git bisect good
62bc04f4c5b4369cc493ff24b1de27aabd79e0c6 is the first bad commit
commit 62bc04f4c5b4369cc493ff24b1de27aabd79e0c6
Author: Werner Lemberg <wl@gnu.org>
Date:   Sat Apr 30 15:37:14 2022 +0200

    * src/sfnt/sfobjs.c (sfnt_load_face): Set `FT_FACE_FLAG_COLOR` for SVG fonts.
    
    Fixes #1151.

 src/sfnt/sfobjs.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)
```
which seems odd, but there you go. I'm not using a Fedora package for SCP, Just the latest `.OTF` release from git.

**Comment by kchibisov** [comment]
@tokyovigilante do you happen to know whether it provides information wrt is the font svg or not? I think that breaks our logic wrt colored fonts which are not scaled, and your particular font has svgs meaning it's treated as colored now?

**Comment by tokyovigilante** [comment]
I would have said no as it is an OTF but not entirely sure how to check. Does seems suspicious.

here's `fc-query` for starters:
```
[Wed 12:27PM] ryan@crackotage ~/$ fc-query ~/.fonts/SourceCodePro-Regular.otf
Pattern has 27 elts (size 32)
	family: "Source Code Pro"(s)
	familylang: "en"(s)
	style: "Regular"(s)
	stylelang: "en"(s)
	fullname: "Source Code Pro"(s)
	fullnamelang: "en"(s)
	slant: 0(i)(s)
	weight: 80(f)(s)
	width: 100(f)(s)
	spacing: 100(i)(s)
	foundry: "ADBO"(s)
	file: "/home/ryan/.fonts/SourceCodePro-Regular.otf"(s)
	index: 0(i)(s)
	outline: True(s)
	scalable: True(s)
	charset:
	0000: 00000000 ffffffff ffffffff 7fffffff 00000000 ffffffff ffffffff ffffffff
	0001: ffffffff ffffffff ffffffff ffffffff 000c8401 00018003 1fffe004 ff300ccc
	0002: 0f000000 00800000 dfff0008 6f7ffdff f337ff9f db8d00d6 7f031fc2 0000001f
	0003: ff0f9fff 3e13d7f9 80000024 44300002 ffffd7f0 fffffffb 2aa37fff 00000002
	0004: ffffffff ffffffff ffffffff 003c000c 0fcf0000 0ccfcc0f 03f38007 000cc3cc
	001d: 00000000 00000000 09c5a388 00000000 10000000 08000001 00000000 00000000
	001e: 00c3f0c0 cffc0ff3 ff0c03ff c000f00f 40ffc03f ffffffff ffffffff 03ffffff
	001f: 00000000 00000000 00000000 00000000 00000000 e0000000 e000e003 6000e000
	0020: 7ffd0080 f62d8047 00000390 e3f30000 001063ff 27365ada 00000000 00000000
	0021: 00c80020 00004045 7fff0000 00000000 03ff0200 00000100 000f0000 00000000
	0022: c626804d 00800a00 00000100 00000033 00000000 00000000 00000000 00000000
	0023: f0010004 00000003 00000000 00000000 00000000 00000000 00000000 00000000
	0025: ffffffff ffffffff ffffffff ffffffff ffffffff f7fc1c03 03009e5f 00000040
	0026: 00230000 1c000000 00000005 00000c69 00000000 00000000 00000000 00000000
	0027: 00080000 00000000 00040000 00000010 00000000 00000000 00000000 000000c0
	002e: 01000000 0000003c 00000000 00000000 00000000 00000000 00000000 00000000
	00e0: 00000000 00000000 00000000 00000000 00000000 000f0007 00000000 00000000
	00fb: 00000006 00000000 00000000 00000000 00000000 00000000 00000000 00000000
	00fe: 0000c000 00000000 00000000 00000000 00000000 00000000 00000000 80000000
	01f3: 00000000 00000000 00000000 00000000 00000000 00600000 00000000 00000000
	01f4: 00000000 00000000 00000000 00000000 00000000 00000200 00000000 00000000
	01f5: 00040000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
	01f9: 00400000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
(s)
	lang: aa|af|ast|av|ay|az-az|ba|be|bg|bi|bin|br|bs|bua|ca|ce|ch|co|cs|cv|cy|da|de|el|en|eo|es|et|eu|fi|fj|fo|fr|fur|fy|gd|gl|gn|gv|haw|ho|hr|hu|ia|ig|id|ie|ik|io|is|it|kaa|ki|kk|kl|kum|kv|ky|la|lb|lez|lt|lv|mg|mh|mi|mk|mo|mt|nb|nds|nl|nn|no|nr|nso|ny|oc|om|os|pl|pt|rm|ro|ru|se|sel|sh|shs|sk|sl|sm|sma|smj|smn|so|sq|sr|ss|st|sv|sw|tg|tk|tl|tn|to|tr|ts|tt|tyv|uk|uz|vi|vo|vot|wa|wen|wo|xh|yap|yo|zu|an|crh|csb|fil|hsb|ht|jv|kj|ku-tr|kwm|lg|li|mn-mn|ms|na|ng|nv|pap-an|pap-aw|qu|quz|rn|rw|sc|sg|sn|su|ty|za(s)
	fontversion: 133038(i)(s)
	capability: "otlayout:DFLT otlayout:cyrl otlayout:grek otlayout:latn"(s)
	fontformat: "CFF"(s)
	decorative: False(s)
	postscriptname: "SourceCodePro-Regular"(s)
	color: False(s)
	symbol: False(s)
	variable: False(s)
	fonthashint: False(s)
	order: 0(i)(s)
```

**Comment by tokyovigilante** [comment]
Hmm, did find this commit in 2020...
https://github.com/adobe-fonts/source-code-pro/commit/4c21bbcb3eec0a2fa2bd77d4356739a414c15c5d
"Updates all font format to remove SVG table."

**Comment by tokyovigilante** [comment]
Ok so just updated my installed versions of SCP to v 2.038, and the bug is gone with freetype master/2.12.1. My previous version was from Jan 2020 so presumably contained an SVG tag and triggered this issue. I guess ultimately a font bug? Would suggest others update their fonts and retest, can probably close this if confirmed.

**Comment by tokyovigilante** [comment]
The Fedora packaged version is also old.
```
Name         : adobe-source-code-pro-fonts
Version      : 2.030.1.050
```

**Comment by rudi10-1** [comment]
Debian does not have Source Code Pro available from package repositories so I have installed it manually from fonts.google.com. Updated my two year old copy and current version works with alacritty. But Fira Code looked quite nice so I'll stick with it for easier maintenance.

**Comment by chrisduerr** [comment]
Okay, so as far as I can tell ultimately SCP doesn't have the SVG bits anymore so people can just update to the latest SCP version to fix this. However any font that does include SVGs will break Alacritty so we should probably try and find a way to work around this.

**Comment by Soveu** [comment]
Hi, could you change the title to "Incorrect cell size with Source Code Pro font after update" or something similar?
I was just about to file another issue, but only after digging into logs and this thread I realized it would be a duplicate.
Thanks

**Comment by kchibisov** [comment]
This is fixed on master, not sure why the issue is around.
